rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,309 @@
+
+The `cluster_maintenance_config` resource will be updated to include a
+`tools_version` field to control the version of client tools advertised by the
+cluster. This resource will not be configurable on Teleport Cloud. It will be
+managed by the Teleport Cloud team.
+
+If cluster administrators want to opt-out of automatic updates and manage
+updates themselves, they will be able to watch the `cluster_maintenance_config`
+resource for changes and push out updates to endpoints manually.
+
+Inspiration drawn from https://go.dev/doc/toolchain.
+
+### Implementation
+
+#### Client tools
+
+##### Automatic updates
+
+The first time a user executes `tsh login` with a version of `tsh` that
+supports automatic updates, the user will be prompted to enroll in automatic
+updates. This choice will be saved in client tools configuration.
+
+```
+$ tsh --proxy=proxy.example.com login
+Keep client tools like tsh and tctl automatically updated? [YES/no]
+
+[...]
+```
+
+```
+$ cat ~/.tsh/config/config.yaml
+enable_autoupdate: true | false
+```
+
+Once enrolled, upon login client tools will check the `tools_version`
+advertised by the cluster at `/v1/webapi/ping`. If this version differs from
+that of the running binary, the recommended version will be downloaded and
+installed.
+
+The original binaries will not be overwritten by automatic updates, instead
+per-cluster binaries with permissions `0555` will be stored at
+`~/.tsh/bin/proxyName/{tctl,tsh}`. This will act as both a cache and allow
+users to connect to different Teleport clusters (running different versions of
+Teleport) without juggling multiple versions of client tools.
+
+```
+$ tree ~/.tsh
+~/.tsh
+├── bin
+│  └── proxy.example.com
+│     ├── tctl
+│     └── tsh
+├── current-profile
+├── keys
+│  └── proxy.example.com
+│     ├── cas
+│     │  └── example.com.pem
+│     ├── certs.pem
+│     ├── foo
+│     ├── foo-ssh
+│     │  └── example.com-cert.pub
+│     ├── foo-x509.pem
+│     └── foo.pub
+├── known_hosts
+└── proxy.example.com.yaml
+```
+
+Users can cancel client tools updates using `Ctrl-C`. This may be needed if the
+user is on a low bandwidth connection (LTE or public Wi-Fi), if the Teleport
+download server is inaccessible, or the user urgently needs to access the
+cluster and can not wait for the update to occur.
+
+```
+$ tsh login --proxy=proxy.example.com
+Client tools are out of date, updating to vX.Y.Z.
+Update progress: [▒▒▒▒▒▒     ] (Ctrl-C to Cancel)
+
+[...]
+```
+
+After downloading, client tools will re-execute the downloaded version.  The
+child process will inherit all environment variables and flags. Re-execution
+will occur only if the parent process is not `tsh` or `tctl`, to prevent
+infinite loops.
+
+An environment variable `TELEPORT_TOOLS_VERSION` will be introduced to enable
+overriding and using a specific version of client tools. This field can be used
+either as an emergency workaround for a known issue or for debugging purposes
```

Comment:
The environment var could support a `TELEPORT_TOOLS_VERSION="local"` value to disable automatic updates without promts. This would be set by default in our docker images for example (in such stateless environments you don't want to DL a new tsh/tctl bin, you just want to bump the whole image).

Reply:
Added support for `TELEPORT_TOOLS_VERSION=off`.